code/modules/explorer_drone/scanner_array.dm

```diff
@@ -0,0 +1,369 @@
+/// Scan condition instances
+GLOBAL_LIST_INIT(scan_conditions,init_scan_conditions())
+
+
+/proc/init_scan_conditions()
+	. = list()
+	for(var/type in subtypesof(/datum/scan_condition))
+		. += new type
+
+#define MAX_SCAN_DISTANCE 10
+
+#define WIDE_SCAN_COST(BAND, SCAN_POWER) (((BAND*BAND)/(SCAN_POWER))*2*60*10)
+#define BASE_POINT_SCAN_TIME 5*60*10
+#define BASE_DEEP_SCAN_TIME 5*60*10
+
+/// Represents scan in progress, only one globally for now, todo later split per z or allow partial dish swarm usage
+/datum/exoscan
+	/// Scan type wide/point/deep
+	var/scan_type
+	/// The scan power this scan was started with, if scanner swarm power falls below this value it will be interrupted
+	var/scan_power = 0
+	/// Target site for point/band scans
+	var/datum/exploration_site/target
+	/// End of scan timer id
+	var/scan_timer
+
+/datum/exoscan/New(scan_type,datum/exploration_site/target)
+	src.scan_type = scan_type
+	src.target = target
+	var/scan_time = 0
+	switch(scan_type)
+		if(EXOSCAN_WIDE)
+			scan_power = length(GLOB.exoscanner_controller.tracked_dishes)
+			scan_time = WIDE_SCAN_COST(GLOB.exoscanner_controller.wide_scan_band,scan_power)
+		if(EXOSCAN_POINT)
+			scan_power = GLOB.exoscanner_controller.get_scan_power(target)
+			scan_time = BASE_POINT_SCAN_TIME/scan_power
+		if(EXOSCAN_DEEP)
+			scan_power = GLOB.exoscanner_controller.get_scan_power(target)
+			scan_time = (BASE_DEEP_SCAN_TIME*target.distance)/scan_power
+	scan_timer = addtimer(CALLBACK(src,.proc/resolve_scan),scan_time,TIMER_STOPPABLE)
+
+/// Short description for in progress scan
+/datum/exoscan/proc/ui_description()
+	switch(scan_type)
+		if(EXOSCAN_WIDE)
+			return "Wide: Scanning sphere starting 1 AU from the station."
+		if(EXOSCAN_POINT)
+			return "Point scan of [target.display_name()]"
+		if(EXOSCAN_DEEP)
+			return "Deep scan of [target.display_name()]"
+
+/datum/exoscan/proc/resolve_scan()
+	switch(scan_type)
+		if(EXOSCAN_WIDE)
+			generate_exploration_sites()
+		if(EXOSCAN_POINT)
+			target.reveal()
+			target.point_scan_complete = TRUE
+		if(EXOSCAN_DEEP)
+			target.reveal()
+			target.deep_scan_complete = TRUE
+	qdel(src)
+
+/datum/exoscan/proc/stop()
+	SEND_SIGNAL(src,COMSIG_EXOSCAN_INTERRUPTED)
+	qdel(src)
+
+/datum/exoscan/Destroy(force, ...)
+	. = ..()
+	deltimer(scan_timer)
+
+/obj/machinery/computer/exoscanner_control
+	name = "Scanner Array Control Console"
+	/// If scan was interrupted show a popup until dismissed.
+	var/failed_popup = FALSE
+	/// Site we're configuring targeted scans for.
+	var/datum/exploration_site/selected_site
+
+/obj/machinery/computer/exoscanner_control/ui_interact(mob/user, datum/tgui/ui)
+	. = ..()
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "ExoscannerConsole", name)
+		ui.open()
+
+/obj/machinery/computer/exoscanner_control/ui_data(mob/user)
+	. = ..()
+	.["failed"] = failed_popup
+	.["selected_site"] = selected_site ? ref(selected_site) : null
```

Comment:
Not necessary, but `selected_site && ref(selected_site)` is a shorter, but still clear way of doing this.